tests/testsuite/check_cfg.rs

```diff
@@ -141,6 +141,53 @@ fn features_with_namespaced_features() {
         .run();
 }
 
+#[cargo_test(nightly, reason = "--check-cfg is unstable")]
+fn features_fingerprint() {
```

Comment:
> This test is verifying the fingerprinting itself but not the end-to-end desired result.

I'm always a bit hesitant to check the diagnostic output of the lint since it changes (sometimes often).

So I just check for the presence of the lint name `unexpected_cfgs` in the output.
And as requested I split-up the PR into a "buggy test commit first" follow by a "impl commit fix".

(I also added a third commit that avoid unnecessary invalidation when reordering the features)